alibi/explainers/cfrl_base.py

```diff
@@ -0,0 +1,1060 @@
+    label matches the target label. Note that this is element-wise, so a tensor is expected to be returned.
+
+    - ``'postprocessing_funcs'``: List[Postprocessing], post-processing list of functions. The function are applied in \
+    the order, from low to high index. Non-differentiable postprocessing can be applied. The function expects as \
+    arguments `x_cf` - the counterfactual instance, `x` - the original input instance and `c` - the conditional \
+    vector, and returns the post-processed counterfactual instance `x_cf_pp` which is passed as `x_cf` for the \
+    following functions. By default, no post-processing is applied (empty list).
+
+    - ``'conditional_func'``: Callable, generates a conditional vector given a input instance. By default, the \
+    function returns `None` which is equivalent to no conditioning.
+
+    - ``'experience_callbacks'``: List[ExperienceCallback], list of callback function applied at the end of each \
+    experience step.
+
+    - ``'train_callbacks'``: List[TrainingCallback], list of callback functions applied at the end of each training \
+    step.
+
+    - ``'actor'``: Optional[keras.Model, torch.nn.Module], actor network.
+
+    - ``'critic;``: Optional[keras.Model, torch.nn.Module], critic network.
+
+    - ``'optimizer_actor'``: Optional[keras.optimizers.Optimizer, torch.optim.Optimizer], actor optimizer.
+
+    - ``'optimizer_critic'``: Optional[keras.optimizer.Optimizer, torch.optim.Optimizer], critic optimizer.
+
+    - ``'lr_actor'``: float, actor learning rate.
+
+    - ``'lr_critic'``: float, critic learning rate.
+
+    - ``'actor_hidden_dim'``: int, actor hidden layer dimension.
+
+    - ``'critic_hidden_dim'``: int, critic hidden layer dimension.
+"""
+
+PARAM_TYPES = {
+    "primitives": [
+        "act_noise", "act_low", "act_high", "replay_buffer_size", "batch_size", "num_workers", "shuffle",
+        "exploration_steps", "update_every", "update_after", "train_steps", "backend", "actor_hidden_dim",
+        "critic_hidden_dim",
+    ],
+    "complex": [
+        "encoder_preprocessor", "decoder_inv_preprocessor", "reward_func", "postprocessing_funcs", "conditional_func",
+        "experience_callbacks", "train_callbacks", "actor", "critic", "optimizer_actor", "optimizer_critic",
+        "encoder", "decoder", "predictor", "sparsity_loss", "consistency_loss",
+    ]
+}
+"""
+Parameter types for serialization
+
+    - ''`primitives`'': List[str], list of parameters having primitive data types.
+
+    - ''`complex`'': List[str], list of parameters having complex data types (e.g., functions, models, optimizers etc.)
+"""
+
+
+class CounterfactualRLBase(Explainer, FitMixin):
+    """
+     Counterfactual Reinforcement Learning Base.
+     TODO: check if all tensors are uppercase
+    """
+
+    def __init__(self,
+                 predictor: Callable,
+                 encoder: Union['keras.Model', 'nn.Module'],
+                 decoder: Union['keras.Model', 'nn.Module'],
```

Comment:
If you check the API docs you will notice that these types are not rendered nicely, I think this is because `keras.Model` and `nn.Module` are only well defined if `TYPE_CHECKING` is true which does not happen when running `sphinx` to build docs. I would recommend using the fully qualified class names, i.e. `tf.keras.Model` and `torch.nn.Module` and check if the problem goes away (there may still be an issue due to `TYPE_CHECKING` being False...).

Reply:
Still not working with `tf.keras.Model` and `torch.nn.Module`.

Reply:
The workaround:
Union['tf.keras.Model', 'torch.nn.Module'] -> 'Union[tf.keras.Model, torch.nn.Module']

Reply:
Check if `mypy` type-checking still works with these hints, e.g. feed a `numpy` array into a function of this input signature and see if an error is raised.